018-cli-2.0/018-cli-2.0.md

```diff
@@ -0,0 +1,431 @@
+    "functions": [
+        {
+          "name" : "My awesome function 💪",
+          "path" : "./FunctionOne",
+          "command" : "python main.py",
+          "runtime": "python3.9",
+          // This section is useful when we allow cloud functions to be tested locally
+          "vars" :{
+            "KEY 1" : "VALUE 1",
+          }
+        },
+    ],
+    "hosting": []
+}
+```
+
+There are two scenarios here 
+* A developer would like to link the directory to an existing project
+
+```sh
+appwrite init
+
+How would you like to start? 
+⭕️ Create a new project
+✅️ Link to an existing project
+
+Choose a project to link this directory to 
+⭕️ Project 1 (nowfoms)
+✅️ Project 2 (dfsdda)
+⭕️ Project 3 (wedfdd)
+```
+
+* A developer would like to create a new project while initializing this directory
+
+```sh
+appwrite init
+
+How would you like to start? 
+✅️ Create a new project
+⭕️ Link to an existing project
+
+Give your project a name:
+Project X
+
+Creating Project X...
+
+Project X Created 👍
+```
+
+Once you choose the project, appwrite.json is initialized with the newly created project's ID?
+
+```json
+{
+  "projectId" : "xyz"
+}
+```
+
+> We need to first add support to utopia-php/CLI to allow for single select and multi select commands to enable this feature.
+
+
+### 🟢 The `appwrite deploy` command
+
+The `appwrite deploy` command is a convenient wrapper around three main commands
+* Create Function
+* Create Tag
+* Update Tag to activate it 
+
+```sh
+# Usage 1
+appwrite deploy functions --confirm 
+
+Which functions would you like to deploy? 
+✅️ FunctionOne
+✅️ FunctionTwo
+⭕️ FunctionThree
+
+Deploying FunctionOne.. 
+FunctionOne already exists 
+Updating FunctionOne with new tag
+Do you want to activate FunctionOne? [Yes/No]
+
+
+# Usage 2
+appwrite deploy functions --all
+# Deploys all the functions specified in appwrite.json
+
+# Future support for static hosting
+appwrite deploy site
+appwrite deploy site --all 
+```
+
+
+## Easier creation of cloud functions
+---
+
+We will introduce an `appwrite init function` command that does the following
+* Generate a function / static site template. 
+* Make an entry in the `appwrite.json` file. Throw an error if the file doesn't exist.
+
+Here are some example usages 
+```sh
+# Usage 1
+appwrite init function --functionId="abc" --name="FunctionOne" --runtime=python-3.9
+# Creates a folder called FunctionOne with a template for a sample python function.
+temp
+├── appwrite.json
+└── FunctionOne
+    ├── .appwrite
+    ├── main.py
+    └── requirements.txt
+
+
+# Usage 2
+appwrite init function --functionId="xyz" --name="FunctionTwo" --runtime=node-16.0
+# Creates a folder called FunctionTwo with a template for a sample node function.
+temp 
+├── appwrite.json
+├── FunctionOne
+│   ├── .appwrite
+│   ├── main.py
+│   └── requirements.txt
+└── FunctionTwo
+    ├── index.js
+    ├── node_modules
+    └── package.json
+
+
+# Usage 3
+appwrite init function
+
+Give your function an ID
```

Comment:
Is it possible to validate this ID with the Key validator as a part of the CLI? We could provide real-time feedback even before a request is made.